File: programmers/다리를 지나는 트럭.cpp
#include <string>
#include <queue>
#include <vector>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    int truckFlag=0;
    int bWeight=0;
    queue<int> q;
    while(bridge_length--){
        q.push(0);
    }
    while(!q.empty()){
        bWeight-=q.front();
        q.pop();
        answer++;
        if(truckFlag==truck_weights.size())
            continue;
        if(bWeight+truck_weights[truckFlag]>weight){
            q.push(0);
        }
        else{
            q.push(truck_weights[truckFlag]);
            bWeight+=truck_weights[truckFlag];
            truckFlag++;
        }
       
    }
    return answer;
}

File: programmers/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> q = new LinkedList<Integer>();
        int answer = 0;
        int bWeight = 0;
        int flag = 0;
        for (int i = 0; i < bridge_length; i++) {
            q.offer(0);
        }
        while (flag < truck_weights.length) {
            bWeight -= q.element();
            q.remove();
            answer++;
            if (bWeight + truck_weights[flag] <= weight) {
                q.offer(truck_weights[flag]);
                bWeight += truck_weights[flag];
                flag++;
            } else {
                q.offer(0);
            }


        }
        answer += bridge_length;

        return answer;
    }
}